source/extensions/common/tap/tap_config_base.cc

```diff
@@ -72,7 +74,15 @@ TapConfigBaseImpl::TapConfigBaseImpl(envoy::config::tap::v3::TapConfig&& proto_c
     NOT_REACHED_GCOVR_EXCL_LINE;
   }
 
-  buildMatcher(proto_config.match_config(), matchers_);
+  envoy::config::common::matcher::v3::MatchPredicate match;
+  if (proto_config.has_match()) {
+    match = proto_config.match();
+  } else {
+    std::string out;
+    ASSERT(proto_config.match_config().SerializeToString(&out));
+    ASSERT(match.ParseFromString(out));
```

Comment:
Is this effectively a wire cast? I don't think this will always be safe as we add new fields, right? Or I guess it will as long as the new proto stays backwards compatible? Can you add comments here? Also see `wireCast` in the version conversion code and other utilities that we have in protobuf/utility.h. I think we do this elsewhere already.

Also, as neither will be required this part needs extra error handling to make sure one of these is set.

Reply:
As long as the MatchPredicate is wire compatible today, and we don't add a new field to the old one and different new field to the new one, we should be good with wireCast.

Reply:
yes, this is doing a wire cast, I changed to use `VersionConverter::upgrade` and also check at least one of them are set since I removed the validation in the proto.

This should work since the whole `proto_config` could have been converted from v2alpha to v3 before calling to this function. Here we do one more conversion from the v3.MatchPredicate to the refactored the factored out v3.MatchPredicate. We will need to maintain the backward compatibility in the factored out MatchPreficate just like any other API.